From f933537c63c6140707b96463f0573bd75f9a3b23 Mon Sep 17 00:00:00 2001
From: Ryosuke Saito <rsaito@redhat.com>
Date: Sun, 15 Nov 2020 22:39:01 +0900
Subject: [PATCH] PCI: imx: Use the external clock as REF_CLK when needed for
 i.MX8MQ

Do not use the external clock when the internal PLL is used as PCIe
REF_CLK.

Signed-off-by: Ryosuke Saito <rsaito@redhat.com>
---
 drivers/pci/controller/dwc/pci-imx6.c | 15 +++++++--------
 1 file changed, 7 insertions(+), 8 deletions(-)

diff --git a/drivers/pci/controller/dwc/pci-imx6.c b/drivers/pci/controller/dwc/pci-imx6.c
index 1eed334db638..16844df2dfab 100644
--- a/drivers/pci/controller/dwc/pci-imx6.c
+++ b/drivers/pci/controller/dwc/pci-imx6.c
@@ -1595,14 +1595,13 @@ static void imx6_pcie_init_phy(struct imx6_pcie *imx6_pcie)
 		break;
 	case IMX8MQ:
 	case IMX8MQ_EP:
-		/*
-		 * TODO: Currently this code assumes external
-		 * oscillator is being used
-		 */
-		regmap_update_bits(imx6_pcie->iomuxc_gpr,
-				   imx6_pcie_grp_offset(imx6_pcie),
-				   IMX8MQ_GPR_PCIE_REF_USE_PAD,
-				   IMX8MQ_GPR_PCIE_REF_USE_PAD);
+		if (imx6_pcie->ext_osc) {
+			/* Use the external oscillator as REF clock */
+			regmap_update_bits(imx6_pcie->iomuxc_gpr,
+					   imx6_pcie_grp_offset(imx6_pcie),
+					   IMX8MQ_GPR_PCIE_REF_USE_PAD,
+					   IMX8MQ_GPR_PCIE_REF_USE_PAD);
+		}
 		/*
 		 * Regarding to the datasheet, the PCIE_VPH is suggested
 		 * to be 1.8V. If the PCIE_VPH is supplied by 3.3V, the
-- 
2.39.2

